### PR TITLE
[inductor][sdpa] Fuse decomposed _safe_softmax SDPA back to native SDPA

### PR DESCRIPTION
**Pull request by etaf**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185574

Root cause
----------
When F.scaled_dot_product_attention falls back to the math path
(_scaled_dot_product_attention_math), inductor's decompositions lower it
into a sequence of primitive ops:

  - the scale is split as sqrt and folded into Q and K via aten.mul.Scalar
  - the two matmuls become aten.bmm (with expand + clone + view around them
    because real Q/K/V are non-contiguous, coming from .transpose(1, 2))
  - the softmax is aten._safe_softmax, which further decomposes into an
    fp32 upcast (convert_element_Root cause
----------
When F.scaled_dot_product_attention falls back to the math path
(_scaled_dot_product_attention_math), inductor's decompositions lower it
into a sequence of primitive ops:

  - the scale is split as sqrt and folded into Q and K via aten.mul.Scalar
  - the two matmuls become aten.bmm (with expand + clone + view around them
    because real Q/K/V are non-contiguous, coming from .transpose(1, 2))
  - the softmax is aten._safe_softmax, which further decomposes into an
    fp32 upcast (convert_element_type), amax/sub/exp/sum/div, a downcast
    back to fp16, and the eq(-inf)/logical_not/any/where masking that
    guards fully-masked rows.

None of the existing SFDP patterns matched this exact decomposition, so
the attention was left as a string of small bmm + pointwise kernels
instead of being fused back into a single scaled_dot_product_attention
call. This causes a measurable performance regression on BERT-like models
on both CUDA and XPU.

Note: the math fallback path is taken even on CUDA (A100) for
BertForMaskedLM because the HuggingFace model constructs a causal mask
via where/expand that does not satisfy the constraints for flash or
efficient attention backends. This was confirmed by inspecting the real
inductor output_code.py from the huggingface.py benchmark (not just
make_fx tracing which uses different shapes/mask construction).

Fix
---
Add SFDP patterns 29 (with attn_mask) and 30 (no attn_mask) that match the
mul.Scalar / bmm / _safe_softmax decomposition and rewrite it back into a
native _scaled_dot_product_attention call, passing scale = inv_scale**2 to
recombine the split scale. The patterns include the leading permute on
Q/K/V so the traced graph contains the same expand -> clone -> view that
real (non-contiguous) inputs produce; without it the serialized pattern
lacks the clone nodes and fails to match. New gp()/gp_bs1() input shapes
([B, S, H, D] = (2, 8, 4, 16) and (1, 8, 4, 16)) are added so the generated
pattern has the right post-permute shapes for both batch>1 (clone present)
and batch==1 (no clone) cases.

Performance
-----------
BertForMaskedLM, inference, fp16, inductor max-autotune-no-cudagraphs,
batch_size=16, seq_len=512:

```
rm -rf /tmp/torchinductor_$USER
python benchmarks/dynamo/huggingface.py --performance --inference \
  --inductor --device={xpu,cuda} --only=BertForMaskedLM \
  --inductor-compile-mode max-autotune-no-cudagraphs --float16
```

| Platform                      | Before | After  | Speedup |
|-------------------------------|--------|--------|---------|
| NVIDIA A100-PCIE-40GB (CUDA)  | 0.883x | 0.963x | +9.1%   |
| Intel Arc B580 (XPU, oneAPI)  | 0.731x | 0.933x | +27.6%  |

Confirmed via the inductor output_code.py on both platforms that all 12
attention layers are fused: scaled_dot_product references go from 0 to
102, and bmm / _safe_softmax occurrences drop to 0.

Test Plan
---------
Added unit tests for patterns 29 and 30 in
test/inductor/test_fused_attention.py (test_sdpa_rewriter_29_{cpu,gpu}
and test_sdpa_rewriter_30_{cpu,gpu}). Tests verify both batch_size>1
(generates clone nodes) and batch_size=1 (no clone) variants.

```
python test/inductor/test_fused_attention.py \
  SDPAPatternRewriterGpuTests.test_sdpa_rewriter_29_gpu \
  SDPAPatternRewriterGpuTests.test_sdpa_rewriter_30_gpu \
  SDPAPatternRewriterCpuTests.test_sdpa_rewriter_29_cpu \
  SDPAPatternRewriterCpuTests.test_sdpa_rewriter_30_cpu
```

Regenerated the serialized patterns:

```
PYTORCH_GEN_PATTERNS=1 python -c \
  "from torch._inductor.fx_passes.fuse_attention import _sfdp_init; _sfdp_init()"
```

Benchmarked on XPU and CUDA (cache cleared first):

```
rm -rf /tmp/torchinductor_$USER
python benchmarks/dynamo/huggingface.py --performance --inference \
  --inductor --device={xpu,cuda} --only=BertForMaskedLM \
  --inductor-compile-mode max-autotune-no-cudagraphs --float16
```

This PR was authored with the assistance of an AI coding assistant.

Co-authored-by: Claude <noreply@anthropic.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by jansel — CHANGES_REQUESTED** [review]
address automated feedback from claude

**Comment by etaf** [comment]
Hi, @jansel : I've addressed the actionable items below.

### Potential Issues

**4. Pattern 29 replacement output shape mismatch — verified, no change needed.**

I double-checked the trailing nodes of the serialized pattern as suggested. The final node is:

```python
_sfdp_pattern_29_half_bs1_inference = CallFunction(aten.view.default, bmm_default_1, Ignored(), _users=0)
```

The trailing `view` reshapes the last `bmm` output (3D `[B*H, S, D]`) back to 4D `[B, H, S, D]`, and `_users=0` confirms it is the pattern's terminal output node — there are no further reshaping nodes the consumer expects. The replacement's `_scaled_dot_product_attention(query.transpose(1, 2), ...)` also returns `[B, H, S, D]`, so the layouts align. Confirmed end-to-end: `_check_common` compares eager vs. compiled output including all post-pattern nodes (passes for `bs>1` and `bs=1`), and the full `BertForMaskedLM` run produced correct results with 102 SDPA fusions and zero `bmm` / `_safe_softmax`.

**5. `check_train=False` — added training coverage.**

Both `_test_sdpa_rewriter_29/30` now run with `check_train=True` for the `batch_size>1` case (XPU backward is auto-skipped by `_check_common`).

**6. No `torch.float` variant — added.**

Both tests now loop over `[torch.half, torch.float]`, exercising the non-`convert_element_type` (no fp32 upcast) path.

### Style / Minor

**7. "on XPU" comment** — Removed "XPU` in _sfdp_pattern_29`.

**8. `gp` naming** — added a comment: `gp = "generic with permute"; gp_bs1 is the batch_size=1 variant (no clone)`.

**9. Registration ordering** — no change (already noted as fine).

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Review by jansel — APPROVED** [review]
